xbmc/threads/SystemClock.cpp

```diff
@@ -14,20 +14,73 @@
 
 namespace XbmcThreads
 {
-  unsigned int SystemClockMillis()
-  {
-    uint64_t now_time;
-    static uint64_t start_time = 0;
-    static bool start_time_set = false;
-
-    now_time = static_cast<uint64_t>(1000 * CurrentHostCounter() / CurrentHostFrequency());
-
-    if (!start_time_set)
-    {
-      start_time = now_time;
-      start_time_set = true;
-    }
-    return (unsigned int)(now_time - start_time);
-  }
-  const unsigned int EndTime::InfiniteValue = std::numeric_limits<unsigned int>::max();
+
+EndTime::EndTime() : m_startTime(std::chrono::system_clock::now())
+{
+}
+
+EndTime::EndTime(unsigned int millisecondsIntoTheFuture)
+  : m_startTime(std::chrono::system_clock::now()),
+    m_totalWaitTime(static_cast<std::chrono::milliseconds>(millisecondsIntoTheFuture))
+{
+}
+
+void EndTime::Set(unsigned int millisecondsIntoTheFuture)
+{
+  m_startTime = std::chrono::system_clock::now();
+  m_totalWaitTime = static_cast<std::chrono::milliseconds>(millisecondsIntoTheFuture);
+}
+
+bool EndTime::IsTimePast() const
+{
+  if (m_totalWaitTime == InfiniteValue)
+    return false;
+
+  if (m_totalWaitTime == std::chrono::milliseconds(0))
+    return true;
+
+  return ((std::chrono::system_clock::now() - m_startTime) >= m_totalWaitTime);
```

Comment:
using a non monotonic clock is a very bad idea. setting the system clock will result in tons of errors.